CODING_STYLE.md

```diff
@@ -0,0 +1,98 @@
+
+Exceptions
+----------
+
+Never use catch-all except: statements, or catch Exception or StandardError, unless you are
+
+   * re-raising the exception, or
+   * creating an isolation point in the program where exceptions are not propagated but are recorded and suppressed instead, such as protecting a thread from crashing by guarding its outermost block.
+
+Python is very tolerant in this regard and except: will really catch everything including misspelled names, sys.exit() calls, Ctrl+C interrupts, unittest failures and all kinds of other exceptions that you simply don’t want to catch.
+
+Versioning
+----------
+
+Modules that inherit from `VersionableInterface` define a `_version` attribute which states their version. This is a tuple of `(MAJOR, MINOR, PATCH)` numbers, which can then be used for Semantic Versioning (where modifications that change the API in a non-backwards compatible way bump the `MAJOR` version (and set the `MINOR` and `PATCH` to 0) and additive changes increase the `MINOR` version (and set the `PATCH` to 0). Changes that have no effect on the external interface (either input or output form) should have their `PATCH` number incremented.  This allows for callers of the interface to determine when changes have happened and whether their code will still work with it.  Volatility carries out these checks through the requirements system, where a plugin can define what requirements it has.
+
+Shared functionality
+--------------------
+
+Within a plugin, there may be functions that are useful to other plugins.  These are created as `classmethod`s so that the plugin can be depended upon by other plugins in their requirements section, without needing to instantiate a whole copy of the plugin.  It is not a staticmethod, because the caller may wish to determine information about the class the method is defined in, and this is not easily accessible for staticmethods.
+A classmethod usually takes a `context` for its first method (and if it requires one, a configuration string for it second).  All other parameters should generally be basic types (such as strings, numbers, etc) so that future work requiring paralellization does not have complex types to have to keep in sync.  In particular, the idea was to ensure only one context was used per method (and each object brings its own context with it, meaning the function signature should not include objects to avoid discrepancies).
+
+Comprehensions
+--------------
+
+Comprehensions are allowed, however multiple for clauses or filter expressions are not permitted. Optimize for readability, not conciseness.
+
+Lambda functions
+----------------
+
+Okay for one-liners. Prefer generator expressions over map() or filter() with a lambda.
+
+Default Arguments
+-----------------
+
+Default arguments are fine, but not with mutable types (because they're consructed once at module load time and can lead to confusion/errors.)
```

Comment:
```suggestion
Default arguments are fine, but not with mutable types (because they're constructed once at module load time and can lead to confusion/errors.)
```